userguide/object_detection/data-preparation.md

```diff
@@ -0,0 +1,134 @@
+(IG02)](https://lear.inrialpes.fr/people/marszalek/data/ig02/). Our goal is to
+make a detector that can spot bikes and cars. Go ahead and click this link and
+download `ig02-v1.0-bikes.zip` and `ig02-v1.0-cars.zip`. In this tutorial, we
+will assume that you have unzipped these files into `~/Downloads/ig02`. That
+is, you should have the following data structure:
+
+```
+~/Downloads/ig02/
+    bike/
+        bike_001.image.png
+        bike_001.mask.0.png
+        bike_001.mask.1.png
+        bike_001.mask.2.png
+        bike_001.mask.3.png
+        bike_002.image.png
+        ...
+    cars/
+        carsgraz_001.image.png
+        carsgraz_001.mask.0.png
+        carsgraz_002.image.png
+        ...
+```
+
+This dataset contains masks marking the exact pixels belonging to each object
+instance. Each image may have several masks to indicate the presence of
+multiple objects. Red pixels indicate object presence and green indicate
+presence but with [occlusion](https://en.wikipedia.org/wiki/Occultation). For
+object detection, we will convert this richly annotated data to bounding boxes.
+Here is a snippet to turn this into an SFrame:
+
+```python
+import turicreate as tc
+import numpy as np
+import glob
+import os
+
+# Change if applicable
+ig02_path = '~/Desktop/ig02'
+
+# Load all images in random order
+raw_sf = tc.image_analysis.load_images(ig02_path, recursive=True,
+                                       random_order=True)
+
+# Split file names so that we can determine what kind of image each row is
+# E.g. bike_005.mask.0.png -> ['bike_005', 'mask']
+info = raw_sf['path'].apply(lambda path: os.path.basename(path).split('.')[:2])
+
+# Rename as 'name', 'type', 'mask_index'
+info = info.unpack().rename({'X.0': 'name', 'X.1': 'type'})
+
+# Add to our main SFrame
+raw_sf = raw_sf.add_columns(info)
+
+# Extract label (e.g. 'bike') from name (e.g. 'bike_003')
+raw_sf['label'] = raw_sf['name'].apply(lambda name: name.split('_')[0])
+
+# Original path no longer needed
+del raw_sf['path']
+
+# Split into images and masks
+sf_images = raw_sf[raw_sf['type'] == 'image']
+sf_masks = raw_sf[raw_sf['type'] == 'mask']
+
+def mask_to_bbox_coordinates(img):
+    """
+    Takes a tc.Image of a mask and returns a dictionary representing bounding
+    box coordinates: e.g. {'x': 100, 'y': 120, 'width': 80, 'height': 120}
+    """
+    mask = img.pixel_data
+    if mask.max() == 0:
+        return None
+    # Take max along both x and y axis, and find first and last non-zero value
+    x0, x1 = np.where(mask.max(0))[0][[0, -1]]
+    y0, y1 = np.where(mask.max(1))[0][[0, -1]]
+
+    return {'x': (x0 + x1) / 2, 'width': (x1 - x0),
+            'y': (y0 + y1) / 2, 'height': (y1 - y0)}
+
+# Convert masks to bounding boxes (drop masks that did not contain bounding box)
+sf_masks['coordinates'] = sf_masks['image'].apply(mask_to_bbox_coordinates)
+
+# There can be empty masks (which returns None), so let's sort them out
```

Comment:
Maybe "so let's get rid of those" rather than "so let's sort them out"

Reply:
Agreed.

All changes have been pushed. Thanks!